### ISSUE TITLE
`strict-boolean-expressions`: disallow string and number only

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#4331]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**
General problem

**What does your suggested rule do?**
Disallow boolean coercion when the type includes `string` or `number`, otherwise allow it.

I believe this is the same as `"strict-boolean-expressions": [true, "allow-undefined-union", "allow-null-union"]`, but in addition to allowing `undefined` and `null`, it should also allow objects/functions. If that's the case, perhaps we just need a new option for `strict-boolean-expressions`?

**List several examples where your rule could be used**
``` ts
declare const age: number;
// Expect error here
if (age) {
  // …
}
```

``` ts
declare const maybeAge: number | undefined;
// Expect error here
if (maybeAge) {
  // …
}
```

``` ts
declare const providedObj: {} | undefined;
declare const defaultObj: {} | undefined;
// Expect no error here, because the type does not include
// string or number, and therefore coercion is intuitive/safe.
const obj = providedObj || defaultObj;
```

**Additional context**
Boolean coercion of `string`s and `number`s is dangerous and can lead to bugs, because it's easy to forget that `''` and `0` are falsy. However, boolean coercion of `undefined` and `null` is much more intuitive (truthy) and therefore safe.

**Comment by rrogowski** [comment]
I'll take a look into this. Any thoughts on the name for the option?

**Comment by OliverJAsh** [comment]
@rrogowski Awesome! Hmm, does `disallow-string-or-number` work?

**Comment by rrogowski** [comment]
@OliverJAsh this seems like an intuitive enough name to me! However, from what I've seen the convention is to start rule options with `allow`. I can't find an example of a rule option starting with `disallow`. Maybe someone else with more experience can weigh in on this.

**Comment by rrogowski** [comment]
@OliverJAsh Just to be clear, is this the kind of case you're talking about?

```ts
declare const value: {} | () => void;
// Error is thrown, but under `disallow-string-or-number` no error would be thrown.
if (value) {
  // ...
}
```

**Comment by OliverJAsh** [comment]
@rrogowski That should be allowed, and so should any other union that doesn't include `string` or `number`.

**Comment by OliverJAsh** [comment]
@rrogowski No, as per the comment in the example code, I would expect no error here.

> I'm having trouble thinking of a case that the combination of these options do not cover that would warrant a new option.

So I'm now trying `"allow-undefined-union", "allow-null-union"`, and you're right, it produces the desired results according to my original examples.

I found this surprising, because `allow-undefined-union` allows `{} | undefined` but disallows `string | undefined` and `number | undefined`. Coincidentally this is what I seem to need, but this particular option's name would suggest both should be allowed, since they both include `undefined` in the union? Should the option name and docs be improved?

Here is an example of something which will error under `"allow-undefined-union", "allow-null-union"`, but I do not want it to error:

``` ts
declare const boolean: boolean;
declare const someObj: {};
// Expect no error here, because the type does not include
// string or number, and therefore coercion is intuitive/safe.
const obj = boolean && someObj;
```

I only want to disallow types which include string or number. All other union types should be allowed.

**Comment by lukewlms** [comment]
@OliverJAsh We need this too (`const obj = boolean && someObj`) - it looks like the use case is addressed by this recent check-in, which will be in the next deployment!

https://github.com/palantir/tslint/issues/3279#issuecomment-453219116

**Comment by adidahiya** [comment]
Going to close this for now given that #3279 has been fixed, which should cover most use cases
